{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ff7022b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(fasta_file):\n",
    "    from Bio import SeqIO\n",
    "    fasta_sequences = SeqIO.parse(open(fasta_file, encoding=\"iso-8859-1\"), 'fasta')\n",
    "    return fasta_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d23ca320",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"metadata/metadata_filtered.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "04a65540",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus name</th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Pango lineage</th>\n",
       "      <th>Clade</th>\n",
       "      <th>Collection date</th>\n",
       "      <th>Submission date</th>\n",
       "      <th>Location</th>\n",
       "      <th>Host</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/USA/CA-HLX-STM-2G38E65FT/2023</td>\n",
       "      <td>EPI_ISL_18260366</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>XBC.1.6.3</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-08-23</td>\n",
       "      <td>2023-09-14</td>\n",
       "      <td>North America / USA / California</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/USA/CA-HLX-STM-F2E2UZDKH/2023</td>\n",
       "      <td>EPI_ISL_18260880</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-08-21</td>\n",
       "      <td>2023-09-14</td>\n",
       "      <td>North America / USA / California</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Spain/AS-HUCA-232339160/2023</td>\n",
       "      <td>EPI_ISL_17949299</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>BA.2</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-06-16</td>\n",
       "      <td>2023-06-29</td>\n",
       "      <td>Europe / Spain / Asturias</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/Japan/PG-586620/2023</td>\n",
       "      <td>EPI_ISL_18888903</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-10-13</td>\n",
       "      <td>2024-02-14</td>\n",
       "      <td>Asia / Japan / Wakayama</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Japan/PG-593636/2023</td>\n",
       "      <td>EPI_ISL_18888929</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>XBC.1.3</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-12-14</td>\n",
       "      <td>2024-02-14</td>\n",
       "      <td>Asia / Japan / Wakayama</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3637</th>\n",
       "      <td>hCoV-19/Israel/ICH-741198878/2023</td>\n",
       "      <td>EPI_ISL_18255299</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-09-04</td>\n",
       "      <td>2023-09-13</td>\n",
       "      <td>Asia / Israel</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3638</th>\n",
       "      <td>hCoV-19/Netherlands/ZH-EMC-8206/2023</td>\n",
       "      <td>EPI_ISL_18255778</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-09-02</td>\n",
       "      <td>2023-09-13</td>\n",
       "      <td>Europe / Netherlands / Zuid-Holland</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639</th>\n",
       "      <td>hCoV-19/USA/TX-HMH-M-132614/2023</td>\n",
       "      <td>EPI_ISL_18256159</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-08-26</td>\n",
       "      <td>2023-09-13</td>\n",
       "      <td>North America / USA / Texas / Houston</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3640</th>\n",
       "      <td>hCoV-19/USA/TX-HMH-M-132276/2023</td>\n",
       "      <td>EPI_ISL_18256302</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-08-23</td>\n",
       "      <td>2023-09-13</td>\n",
       "      <td>North America / USA / Texas / Houston</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3641</th>\n",
       "      <td>hCoV-19/Croatia/HZJZ_5054/2023</td>\n",
       "      <td>EPI_ISL_18257810</td>\n",
       "      <td>Former VOC Omicron GRA (B.1.1.529+BA.*) first ...</td>\n",
       "      <td>HW.1.1</td>\n",
       "      <td>GRA</td>\n",
       "      <td>2023-08-18</td>\n",
       "      <td>2023-09-14</td>\n",
       "      <td>Europe / Croatia / Split-Dalmatia County</td>\n",
       "      <td>Human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3642 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Virus name      Accession ID  \\\n",
       "0     hCoV-19/USA/CA-HLX-STM-2G38E65FT/2023  EPI_ISL_18260366   \n",
       "1     hCoV-19/USA/CA-HLX-STM-F2E2UZDKH/2023  EPI_ISL_18260880   \n",
       "2      hCoV-19/Spain/AS-HUCA-232339160/2023  EPI_ISL_17949299   \n",
       "3              hCoV-19/Japan/PG-586620/2023  EPI_ISL_18888903   \n",
       "4              hCoV-19/Japan/PG-593636/2023  EPI_ISL_18888929   \n",
       "...                                     ...               ...   \n",
       "3637      hCoV-19/Israel/ICH-741198878/2023  EPI_ISL_18255299   \n",
       "3638   hCoV-19/Netherlands/ZH-EMC-8206/2023  EPI_ISL_18255778   \n",
       "3639       hCoV-19/USA/TX-HMH-M-132614/2023  EPI_ISL_18256159   \n",
       "3640       hCoV-19/USA/TX-HMH-M-132276/2023  EPI_ISL_18256302   \n",
       "3641         hCoV-19/Croatia/HZJZ_5054/2023  EPI_ISL_18257810   \n",
       "\n",
       "                                                Variant Pango lineage Clade  \\\n",
       "0     Former VOC Omicron GRA (B.1.1.529+BA.*) first ...     XBC.1.6.3   GRA   \n",
       "1     Former VOC Omicron GRA (B.1.1.529+BA.*) first ...        HW.1.1   GRA   \n",
       "2     Former VOC Omicron GRA (B.1.1.529+BA.*) first ...          BA.2   GRA   \n",
       "3     Former VOC Omicron GRA (B.1.1.529+BA.*) first ...          HW.1   GRA   \n",
       "4     Former VOC Omicron GRA (B.1.1.529+BA.*) first ...       XBC.1.3   GRA   \n",
       "...                                                 ...           ...   ...   \n",
       "3637  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...        HW.1.1   GRA   \n",
       "3638  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...        HW.1.1   GRA   \n",
       "3639  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...        HW.1.1   GRA   \n",
       "3640  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...        HW.1.1   GRA   \n",
       "3641  Former VOC Omicron GRA (B.1.1.529+BA.*) first ...        HW.1.1   GRA   \n",
       "\n",
       "     Collection date Submission date  \\\n",
       "0         2023-08-23      2023-09-14   \n",
       "1         2023-08-21      2023-09-14   \n",
       "2         2023-06-16      2023-06-29   \n",
       "3         2023-10-13      2024-02-14   \n",
       "4         2023-12-14      2024-02-14   \n",
       "...              ...             ...   \n",
       "3637      2023-09-04      2023-09-13   \n",
       "3638      2023-09-02      2023-09-13   \n",
       "3639      2023-08-26      2023-09-13   \n",
       "3640      2023-08-23      2023-09-13   \n",
       "3641      2023-08-18      2023-09-14   \n",
       "\n",
       "                                      Location   Host  \n",
       "0             North America / USA / California  Human  \n",
       "1             North America / USA / California  Human  \n",
       "2                    Europe / Spain / Asturias  Human  \n",
       "3                      Asia / Japan / Wakayama  Human  \n",
       "4                      Asia / Japan / Wakayama  Human  \n",
       "...                                        ...    ...  \n",
       "3637                             Asia / Israel  Human  \n",
       "3638       Europe / Netherlands / Zuid-Holland  Human  \n",
       "3639     North America / USA / Texas / Houston  Human  \n",
       "3640     North America / USA / Texas / Houston  Human  \n",
       "3641  Europe / Croatia / Split-Dalmatia County  Human  \n",
       "\n",
       "[3642 rows x 9 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omic_acc = 'Former VOC Omicron GRA (B.1.1.529+BA.*) first detected in Botswana/Hong Kong/South Africa'\n",
    "\n",
    "# get accession ids for a specific variant\n",
    "df_omic = df[df['Variant'] == omic_acc]\n",
    "df_omic = df_omic[df_omic['Accession ID'].notnull()]\n",
    "# keep rows only with collection date is newer than 2023-06-01\n",
    "df_omic_new = df_omic[df_omic['Collection date'] > '2023-05-20']\n",
    "# keep only the collection data is in yyyy-mm-dd format\n",
    "df_omic_new = df_omic_new[df_omic_new['Collection date'].str.match(r'\\d{4}-\\d{2}-\\d{2}')]\n",
    "df_omic_new = df_omic_new.reset_index(drop=True)\n",
    "df_omic_new"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "016c89ab",
   "metadata": {},
   "source": [
    "# VARIANT SELECTION (Omicron) Takes around 7 minutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "03acc64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fseq_acc_filtered = read_fasta('spikeprot0508/spikeprot0508_acc_filtered.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6c4edb36",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11445408it [06:45, 28204.75it/s]\n"
     ]
    }
   ],
   "source": [
    "# select the sequences with accession ids in the list\n",
    "omic_new_accs = df_omic_new['Accession ID'].tolist()\n",
    "fseq_omic = []\n",
    "\n",
    "# show progress\n",
    "from tqdm import tqdm\n",
    "for record in tqdm(fseq_acc_filtered):\n",
    "    if record.id in omic_new_accs:\n",
    "        fseq_omic.append(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5f1365b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1903\n"
     ]
    }
   ],
   "source": [
    "print(len(fseq_omic))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "24497aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the spike protein sequences for a specific variant to a fasta file\n",
    "with open('spikeprot0508/spikeprot0508_acc_filtered_omicron_new.fasta', 'w') as f:\n",
    "    for record in fseq_omic:\n",
    "        f.write('>' + record.id + '\\n')\n",
    "        f.write(str(record.seq) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0d578d5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "548\n"
     ]
    }
   ],
   "source": [
    "# get unique sequences for a specific variant\n",
    "def get_unique_sequences(fasta_file):\n",
    "    from Bio import SeqIO\n",
    "    from tqdm.notebook import tqdm\n",
    "    unique_sequences = []\n",
    "    unique_ids = []\n",
    "    for record in SeqIO.parse(open(fasta_file, encoding=\"iso-8859-1\"), 'fasta'):\n",
    "        if str(record.seq) not in unique_sequences:\n",
    "            unique_sequences.append(str(record.seq))\n",
    "            unique_ids.append(record.id)\n",
    "    return unique_sequences, unique_ids\n",
    "\n",
    "unique_sequences, unique_ids = get_unique_sequences('spikeprot0508/spikeprot0508_acc_filtered_omicron_new.fasta')\n",
    "\n",
    "print(len(unique_sequences))\n",
    "\n",
    "# write the unique sequences to a fasta file\n",
    "with open('spikeprot0508/spikeprot0508_acc_filtered_omicron_new_unique.fasta', 'w') as f:\n",
    "    for i, seq in enumerate(unique_sequences):\n",
    "        f.write('>' + unique_ids[i] + '\\n')\n",
    "        f.write(seq + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "21e0e54d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_17793872</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EPI_ISL_17796983</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNFRTGTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPI_ISL_17803121</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EPI_ISL_17803165</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPI_ISL_17803190</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITRTQSYTNSFTRGVYYPDKVFRSSVLH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>EPI_ISL_19084465</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITTTQXXXAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>EPI_ISL_19084464</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITTTQXXXXYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>EPI_ISL_19084596</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>EPI_ISL_18229220</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>EPI_ISL_19086907</td>\n",
       "      <td>MFVFLVLLPLVSSQCVMPLFNLITTTQSYTNSFTRGVYYPDKVFRS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>548 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         accession_id                                           sequence\n",
       "0    EPI_ISL_17793872  MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...\n",
       "1    EPI_ISL_17796983  MFVFLVLLPLVSSQCVNFRTGTQLPPAYTNSFTRGVYYPDKVFRSS...\n",
       "2    EPI_ISL_17803121  MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...\n",
       "3    EPI_ISL_17803165  MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...\n",
       "4    EPI_ISL_17803190  MFVFLVLLPLVSSQCVNLITRTQSYTNSFTRGVYYPDKVFRSSVLH...\n",
       "..                ...                                                ...\n",
       "543  EPI_ISL_19084465  MFVFLVLLPLVSSQCVNLITTTQXXXAYTNSFTRGVYYPDKVFRSS...\n",
       "544  EPI_ISL_19084464  MFVFLVLLPLVSSQCVNLITTTQXXXXYTNSFTRGVYYPDKVFRSS...\n",
       "545  EPI_ISL_19084596  MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...\n",
       "546  EPI_ISL_18229220  MFVFLVLLPLVSSQCVNLITRTQLSPAYTNSFTRGVYYPDKVFRSS...\n",
       "547  EPI_ISL_19086907  MFVFLVLLPLVSSQCVMPLFNLITTTQSYTNSFTRGVYYPDKVFRS...\n",
       "\n",
       "[548 rows x 2 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# write to a csv file\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "def write_to_csv(fasta_file, csv_file):\n",
    "    with open(csv_file, 'w') as f:\n",
    "        f.write('accession_id,sequence\\n')\n",
    "        for record in SeqIO.parse(open(fasta_file, encoding=\"iso-8859-1\"), 'fasta'):\n",
    "            f.write(record.id + ',' + str(record.seq) + '\\n')\n",
    "\n",
    "write_to_csv('spikeprot0508/spikeprot0508_acc_filtered_omicron_new_unique.fasta', 'spikeprot0508/unique_Omicron_548.csv')\n",
    "\n",
    "# read the csv file\n",
    "unique_Omicron_548 = pd.read_csv('spikeprot0508/unique_Omicron_548.csv')\n",
    "unique_Omicron_548\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "008735ae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch38",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
